File: src/test/java/io/tripled/social/client/presentation/request/CliRequestFactoryTest.java
package io.tripled.social.client.presentation.request;

import io.tripled.social.client.presentation.CliRequest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.instanceOf;

class CliRequestFactoryTest {

  private CliRequestFactory factory;

  @BeforeEach
  void setUp() throws Exception {
    factory = new CliRequestFactory();
  }

  @Test
  void returnsPostCommand() throws Exception {
    String input = "Alice -> Hi there!";

    CliRequest request = factory.create(input).orElse(null);

    assertThat(request, instanceOf(PostCliRequest.class));
  }

  @Test
  void returnsPostCommandWithTrimmedValues() throws Exception {
    String input = "  Alice   ->   Hi There!  ";

    CliRequest request = factory.create(input).orElse(null);

    assertThat(request, instanceOf(PostCliRequest.class));
    assertThat(request, equalTo(new PostCliRequest("Alice", "Hi There!")));
  }

  @Test
  void returnsReadCommand() throws Exception {
    String input = "Alice";
    String message = "  Alice  ";

    CliRequest request = factory.create(input).orElse(null);

    assertThat(request, instanceOf(ReadCliRequest.class));
  }

  @Test
  void returnsReadCommandWithTrimmedValues() throws Exception {
    String input = "  Alice  ";

    CliRequest request = factory.create(input).orElse(null);

    assertThat(request, instanceOf(ReadCliRequest.class));
    assertThat(request, equalTo(new ReadCliRequest("Alice")));
  }

  @Test
  void returnsFollowCommand() throws Exception {
    String input = "Alice follows Ben";

    CliRequest cliRequest = factory.create(input).orElse(null);

    assertThat(cliRequest, instanceOf(FollowCliRequest.class));
  }

  @Test
  void retunsWallCommand() throws Exception {
    String input = "Ben wall";

    CliRequest cliRequest = factory.create(input).orElse(null);

    assertThat(cliRequest, instanceOf(WallCliRequest.class));
  }
}

File: src/test/kotlin/io/tripled/social/client/presentation/request/CliRequestFactoryTest.kt
package io.tripled.social.client.presentation.request

import io.tripled.social.client.presentation.CliRequest
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class CliRequestFactoryTest {
    private val factory: CliRequestFactory = CliRequestFactory()

    @Test
    @Throws(Exception::class)
    fun returnsPostCommand() {
        val input = "Alice -> Hi there!"

        val request: CliRequest? = factory.create(input)

        MatcherAssert.assertThat<CliRequest?>(request, Matchers.instanceOf<CliRequest?>(PostCliRequest::class.java))
    }

    @Test
    @Throws(Exception::class)
    fun returnsPostCommandWithTrimmedValues() {
        val input = "  Alice   ->   Hi There!  "

        val request: CliRequest? = factory.create(input)

        MatcherAssert.assertThat<CliRequest?>(request, Matchers.instanceOf<CliRequest?>(PostCliRequest::class.java))
        MatcherAssert.assertThat<CliRequest?>(
            request,
            Matchers.equalTo<CliRequest?>(PostCliRequest("Alice", "Hi There!"))
        )
    }

    @Test
    @Throws(Exception::class)
    fun returnsReadCommand() {
        val input = "Alice"
        val message = "  Alice  "

        val request: CliRequest? = factory.create(input)

        MatcherAssert.assertThat<CliRequest?>(request, Matchers.instanceOf<CliRequest?>(ReadCliRequest::class.java))
    }

    @Test
    @Throws(Exception::class)
    fun returnsReadCommandWithTrimmedValues() {
        val input = "  Alice  "

        val request: CliRequest? = factory.create(input)

        MatcherAssert.assertThat<CliRequest?>(request, Matchers.instanceOf<CliRequest?>(ReadCliRequest::class.java))
        MatcherAssert.assertThat<CliRequest?>(request, Matchers.equalTo<CliRequest?>(ReadCliRequest("Alice")))
    }

    @Test
    @Throws(Exception::class)
    fun returnsFollowCommand() {
        val input = "Alice follows Ben"

        val cliRequest: CliRequest? = factory.create(input)

        MatcherAssert.assertThat<CliRequest?>(
            cliRequest,
            Matchers.instanceOf<CliRequest?>(FollowCliRequest::class.java)
        )
    }

    @Test
    @Throws(Exception::class)
    fun retunsWallCommand() {
        val input = "Ben wall"

        val cliRequest: CliRequest? = factory.create(input)

        MatcherAssert.assertThat<CliRequest?>(cliRequest, Matchers.instanceOf<CliRequest?>(WallCliRequest::class.java))
    }
}